src/components/TranslationDialog/TranslationModal/TranslationForm.js

```diff
@@ -0,0 +1,216 @@
+import { useAlert, useDataMutation } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import {
+    CenteredContent,
+    DataTable,
+    DataTableBody,
+    DataTableCell,
+    DataTableColumnHeader,
+    DataTableHead,
+    DataTableRow,
+    InputField,
+    ModalContent,
+} from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React, { useEffect, useMemo, useState } from 'react'
+import { LocalesSelect } from './LocalesSelect.js'
+import { TranslationModalActions } from './TranslationModalActions.js'
+
+const SESSION_STORAGE_TRANSLATION_LOCALE_KEY =
+    'translation-dialog-selected-locale'
+
+const getTranslationsMutation = resource => ({
+    resource: `${resource}/translations`,
+    type: 'update',
+    data: ({ translations }) => ({ translations }),
+})
+
+export const TranslationForm = ({
+    fieldsToTranslate,
+    objectToTranslate,
+    translations,
+    resource,
+    onTranslationSaved,
+    onClose,
+}) => {
+    const [newTranslations, setNewTranslations] = useState()
+    const [translationLocale, setTranslationLocale] = useState()
+
+    const { show: showError } = useAlert(error => error, { critical: true })
+
+    const formatFieldLabel = field => {
+        field
+            .replace(/[a-z][A-Z]/g, match =>
+                [match.charAt(0), match.charAt(1)].join(' ')
+            )
+            .toLowerCase()
+
+        return field.charAt(0).toUpperCase() + field.slice(1)
+    }
+
+    const camelCaseToUnderscores = field =>
+        field
+            .replace(/[a-z][A-Z]/g, match =>
+                [match.charAt(0), match.charAt(1)].join('_')
+            )
+            .toLowerCase()
+
+    const getTranslationIndexForField = field =>
+        newTranslations.findIndex(
+            element =>
+                element.locale === translationLocale &&
+                element.property.toLowerCase() === camelCaseToUnderscores(field)
+        )
+
+    const getTranslationForField = field => {
+        const translationIndex = getTranslationIndexForField(field)
+
+        return translationIndex !== -1
+            ? newTranslations[translationIndex]?.value || ''
+            : ''
+    }
+
+    const setTranslationForField = (field, translation) => {
+        const newTranslation = {
+            locale: translationLocale,
+            property: camelCaseToUnderscores(field).toUpperCase(),
+            value: translation,
+        }
+
+        const translationIndex = getTranslationIndexForField(field)
+
+        const updatedTranslations = [...newTranslations]
+
+        if (translationIndex !== -1) {
+            updatedTranslations.splice(translationIndex, 1, newTranslation)
+        } else {
+            updatedTranslations.push(newTranslation)
+        }
+
+        setNewTranslations(updatedTranslations)
+    }
+
+    const translationsMutation = useMemo(
+        () => getTranslationsMutation(resource),
+        []
+    )
```

Comment:
We should not use `useMemo` for this [according to the docs](https://reactjs.org/docs/hooks-reference.html#usememo): 
> **You may rely on useMemo as a performance optimization, not as a semantic guarantee.**
> In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without `useMemo` — and then add it to optimize performance.

As we've done in the events-reports app, we should use `useRef` instead.

Reply:
👍 
Done.